Plugin initial implementation connect ExpressGateway/express-gateway#392

XVincentX opened the pull request:
This PR implements a rewrite plugin based on RegExp and Express Paths.

It contains 2 `conditions` and one `policy` based on multiple discussion we had during the last days.

The implementation has been partially influenced by `http-rewrite-middleware` - kudos to these guys.

- [x] Tests

DrMegavolt commented on .vscode/launch.json:
```diff
@@ -0,0 +1,16 @@
+{
```
can you add  gitignore with standard items like .idea , .vscode etc

XVincentX replied:
Definitely!

DrMegavolt commented on README.md:
```diff
@@ -0,0 +1,59 @@
+          match: /tina/:code
+        action:
+          rewrite: /status/:code
+          redirect: permanent
```
Why not have 301,302 , 307 etc here ?

XVincentX replied:
Good point - I honestly ignored all the other status codes as they felt "unused" to me. I'll have a quick look to the spec again and modify the code accordingly.

DrMegavolt commented on manifest.js:
```diff
@@ -0,0 +1,66 @@
+            return next();
+          }
+
+          res.statusCode = actionParams.redirect === 'permanent' ? 301 : 302;
```
see comment above, 301,302 are not the only options here

DrMegavolt commented on manifest.js:
```diff
@@ -0,0 +1,66 @@
+
+          res.statusCode = actionParams.redirect === 'permanent' ? 301 : 302;
+          res.setHeader('Location', toUrl);
+          res.end();
```
why not use res.redirect(301,toUrl); ?

XVincentX replied:
Good point, I think it's handling more cases than I could imagine. I'll fix it.

DrMegavolt commented on test/server.js:
```diff
@@ -0,0 +1,6 @@
+
+gateway()
+  .load(path.join(__dirname, 'config'))
+  .run();
```
I don't see any test here except it runs. 
Minimal variant is to run real server + fake backend server and ensure that fake backend is recieving correct url
same for redirect

XVincentX replied:
Definitely - as I wrote in the pull request text, I'm figuring out how to test but more importantly - I'm missing a way to shut down the gateway programmatically. I can see from the gateway repo you guys have an utility to fork and kill the process and I'm not entirely sure whether I should just copy paste that code here or eventually adding a shutdown function to the gateway. What's your opinion?

XVincentX replied:
Addressed!

DrMegavolt commented on test/config/gateway.config.yml:
```diff
@@ -0,0 +1,39 @@
+          -
+            condition:
+              name: pathmatch
+              match: /gigino/:code
```
Could you please check how variant with multi-segment match is working 
match /api/:route*
rewrite /api/v1/:route

incoming url  /api/users/nick
proxied to /api/v1/users/nick

XVincentX replied:
👍 I'll code all these as test cases.

XVincentX commented:
@DrMegavolt I've addressed all your comments as well added several tests to make sure the gateway is interacting correctly with the plugin.

Let me know if you want me to go further and add unit tests for the functions as well - although I do not see any great advantage here as most of the work is done by the internal dependency.

stanislaw-glogowski commented on plugin.test.js:
```diff
@@ -0,0 +1,69 @@
+let Application = undefined;
+
+beforeAll((done) => {
+  axios.defaults.baseURL = 'http://localhost:8080/';
```
You should consider using `axios.create` instead of changing configuration of default instance of axios.

XVincentX replied:
@stanislaw-glogowski Done
